ouroboros-consensus/src/Ouroboros/Consensus/Protocol/PBFT.hs

```diff
@@ -214,26 +222,101 @@ instance (PBftCrypto c, Typeable c) => OuroborosTag (PBft c) where
       case Bimap.lookupR (hashVerKey pbftIssuer) dms of
         Nothing -> throwError $ PBftNotGenesisDelegate (hashVerKey pbftIssuer) lv
         Just gk -> do
-          let totalSigners = Seq.length signers
-              gkSigners = Seq.length (Seq.filter (== gk) signers)
+          let totalSigners = chainStateSize chainState
+              gkSigners = maybe 0 Seq.length $ Map.lookup gk chainState
           when (totalSigners >= winSize && gkSigners > wt)
             $ throwError (PBftExceededSignThreshold totalSigners gkSigners)
-          return $! takeR (winSize + 2*k) chainState Seq.|> (gk, blockSlot b)
+          return $! insertSigner gk (blockSlot b) $ pruneChainState (winSize + 2*k) chainState
     where
       PBftParams{..} = pbftParams
       PBftFields{..} = headerPBftFields cfg b
       winSize = fromIntegral pbftSignatureWindow
       SecurityParam (fromIntegral -> k) = pbftSecurityParam
       wt = floor $ pbftSignatureThreshold * fromIntegral winSize
-      -- Take the rightmost n elements of a sequence
-      takeR :: Integral i => i -> Seq a -> Seq a
-      takeR (fromIntegral -> n) s = Seq.drop (Seq.length s - n - 1) s
 
   rewindChainState _ cs mSlot = case mSlot of
-    Origin  -> Just Seq.empty
-    At slot -> case Seq.takeWhileL (\(_, s) -> s <= slot) cs of
-        _ Seq.:<| _ -> Just cs
-        _           -> Nothing
+    Origin  -> Just Map.empty
+    At slot -> let oldCs = Map.map (Seq.takeWhileL (\s -> s <= slot)) cs in
+      if (any (not . Seq.null) $ Map.elems oldCs) then Just oldCs else Nothing
+
+-- | Prune the chain state to the given size by dropping the signers in the
+-- oldest slots.
+--
+-- Examples
+--
+-- $setup
+-- >>> :set -XOverloadedLists
+-- >>> :set -XScopedTypeVariables
+-- >>> import Test.QuickCheck
+-- >>> :{
+--   newtype TestChainState = TestChainState (Map Int (Seq Int))
+--   instance Arbitrary TestChainState where
+--     arbitrary = do
+--       numElts <- arbitrary
+--       ub <- arbitrary
+--       return . TestChainState $ Map.fromList
+--         [(idx, elts) | idx <- [0..numElts]
+--                      , let elts = Seq.fromList [i | i <- [0..ub], i `mod` numElts == idx]
+--         ]
+-- :}
+
+-- prop> \cs -> pruneChainState (chainStateSize cs) cs == cs
+--
+-- prop> \n (TestChainState cs) -> n > 0 && (chainStateSize cs) > n ==> (chainStateSize (pruneChainState n cs)) == n
+--
+-- >>> :{
+--   let (cs :: Map Char (Seq Int)) = [('a', [1,2,5]), ('b', [3, 4, 6])]
+--   in pruneChainState 2 cs
+-- :}
+-- fromList [('a',fromList [5]),('b',fromList [6])]
+--
+pruneChainState :: forall k v. (Show k, Show v, Ord k, Ord v) => Int -> Map k (Seq v) -> Map k (Seq v)
```

Comment:
Why do we need the `Show` constraints?

Reply:
We don't, they were there during my testing, forgot to take them out...